### PR TITLE
[PLUGIN-1718] Allow flexible column names (Japanese Characters)

### PR DESCRIPTION
**Pull request by psainics**
## Allow flexible column names (Japanese Characters)

Jira : [PLUGIN-1718](https://cdap.atlassian.net/browse/PLUGIN-1718)

Let the user enter non English characters as column names.

A column name can contain the letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with a letter or underscore. For more flexible column name support, see [flexible column names](https://cloud.google.com/bigquery/docs/schemas#flexible-column-names). 

### Code change 
- Some classes were moved from a dependency as it was not being maintained and had old regex, these classes are present under the lib package.
- Updated Docs for  BQ and BQMT 
- Update column name regex 
- Update when to use JSON Writer
  - If the field name is not in English characters, then we will use JSON format, We do this as the AVRO load job in BQ does not support non-English characters in field names for now 
- Added new unit test case
- Some classes were moved from com.google.cloud.hadoop.io.bigquery bigquery-connector library
- Update the table regex as per the requirement of flexible column names
- All moved classes are under lib folder
  
 ### Unit Test
 
 - testValidateColumnName`WithSpecialCharacter`
  - testValidateColumnName`WithNumbers`
  - testValidateColumnName`WithCapitalLetters`
  - testValidateColumnName`WithDash`
  - testValidateColumnName`WithUnderscore`
  - testValidateColumnName`WithEmoji`
  - testValidateColumnName`WithSpace`
  - testValidateColumnName`WithJapaneseColumnName`
  - testValidateColumnName`WithInvalidColumnName`
  - testValidateColumnName`WithChineseColumnName`
  - testValidateColumnName`WithValidColumnName` 
  - testValidateColumnName`WithChineseColumnName`
  - testValidateColumnName`WithInvalidColumnName`
  - testValidateColumnName`WithJapaneseColumnName`
  - testValidateColumnName`WithSpace`
  - testValidateColumnName`WithEmoji`
  - testValidateColumnName`WithUnderscore`
  - testValidateColumnName`WithDash`
  - testValidateColumnName`WithCapitalLetters`
  - testValidateColumnName`WithNumbers`
  - testValidateColumnName`WithSpecialCharacter`
  - testValidateColumnName`With300Length`
  - testValidateColumnName`With301Length`
  
image

[PLUGIN-1718]: https://cdap.atlassian.net/browse/PLUGIN-1718?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java [hunk 1]
```diff
@@ -75,6 +76,7 @@ public final class BigQuerySinkConfig extends AbstractBigQuerySinkConfig {
   public static final String NAME_RANGE_INTERVAL = "rangeInterval";
 
   public static final int MAX_NUMBER_OF_COLUMNS = 4;
+  private static final int MAX_LENGTH_OF_COLUMN_NAME = 300;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java:80]: Add a link to documentation that specifies max column name length - https://cloud.google.com/bigquery/docs/schemas#column_names

File: src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexBatchSink.java [hunk 2]
```diff
@@ -89,6 +88,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexBatchSink.java]: nit: remove extra newline

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java [hunk 3]
```diff
@@ -82,7 +82,7 @@ public final class BigQueryUtil {
 
   public static final String BUCKET_PATTERN = "[a-z0-9._-]+";
   public static final String DATASET_PATTERN = "[A-Za-z0-9_]+";
-  public static final String TABLE_PATTERN = "[A-Za-z0-9_-]+";
+  public static final String TABLE_PATTERN = "[\\p{L}\\p{M}\\p{N}\\p{Pc}\\p{Pd}\\p{Zs}]+";
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java:86]: can you link to documentation that indicates that table name can also include characters in other languages? Also, if this is supported can we add unit tests for table names in other languages?

**Review by rmstar — COMMENTED** [review]
> Some classes were moved from a dependency as it was not being maintained and had old regex, these classes are present under the lib package.


Can you please clarify why we couldn't just upgrade to a newer version of the `com.google.cloud.bigdataoss:bigquery-connector` that has support for flexible column names?

**Comment by psainics** [comment]
> > Some classes were moved from a dependency as it was not being maintained and had old regex, these classes are present under the lib package.
> 
> Can you please clarify why we couldn't just upgrade to a newer version of the `com.google.cloud.bigdataoss:bigquery-connector` that has support for flexible column names?

Certainly, the reason we couldn't upgrade to a newer version of com.google.cloud.bigdataoss:bigquery-connector is because the library isn't actively maintained. The decision to move classes was based on this, and additionally, new Spark libraries are being recommended. Integrating those is a possibility for the future, but for now, this adjustment serves as a quick fix.

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java (hunk 3 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java:86]: Added a comment
> // Read more about table naming in BigQuery here: https://cloud.google.com/bigquery/docs/tables#table_naming

Added new test cases for valid and invalid table names, including international characters.

File: src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexBatchSink.java (hunk 2 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexBatchSink.java]: Removed !

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java (hunk 1 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java:80]: Added a comment 
> // As defined in https://cloud.google.com/bigquery/docs/schemas#column_names

**Comment by rmstar** [comment]
> Certainly, the reason we couldn't upgrade to a newer version of com.google.cloud.bigdataoss:bigquery-connector is because the library isn't actively maintained. The decision to move classes was based on this, and additionally, new Spark libraries are being recommended. Integrating those is a possibility for the future, but for now, this adjustment serves as a quick fix.

I see, thanks for the context. 

Looking at https://cloud.google.com/dataproc/docs/tutorials/bigquery-connector-mapreduce-example, it looks like the BQ connector is not actively maintained like you said, and the replacement is [Spark BigQuery connector](https://github.com/GoogleCloudDataproc/spark-bigquery-connector). Can you please open a Jira to migrate to this connector and add a comment in the copy pasted code indicating that we should migrate the Spark BQ connector.
Something like this:
```
// TODO(PLUGIN-XXXX): Migrate to Spark BigQuery connector. 
```

**Review by rmstar — APPROVED** [review]

**Pull request merged**